document/js-api/index.bs

```diff
@@ -801,9 +822,37 @@ Assert: |type| is not [=𝗂𝟨𝟦=].
 1. If |type| is [=𝖿𝟨𝟦=],
     1. Let |f64| be ? [=ToNumber=](|v|).
     1. Return [=𝖿𝟨𝟦.𝖼𝗈𝗇𝗌𝗍=] |f64|.
+1. If |type| is [=anyref=],
+    1. If |v| is a primitive value but not a string, symbol, or null, throw |error|.
```

Comment:
What's the criteria for allowing Strings here (which I think get implicitly boxed), while primitive number/boolean are disallowed although they could also be implicitly boxed?

Reply:
Excellent question, which has also been asked by @gahaas. I believe it would be rather useful to allow any JS value here, but @lukewagner mentioned some reasons why that might be difficult in SpiderMonkey. I have it on my slides as an open question for this week's meeting.

Reply:
Well, my expectation here is that string promitives wouldn't autobox into String objects.  So, e.g., the identity function would preserve primitive-ness.

Reply:
The problem is the implementation of ref.eq. As far as I understand it, non-primitive JS objects have an identity, so we can implement ref.eq by comparing the references to two JS objects, instead of doing a deep comparison. Primitive values, however, don't have an identity. The same number, or the same string, may be stored multiple times on the heap, in different representations. Therefore we would have to do a deep comparison for primitive values, or we would have to canonicalize primitive values in ToWebAssemblyValue to allow a reference comparison again.

Reply:
@gahaas, ref.eq requires `anyeqref`, and strings are excluded from that (for the reasons you describe). So would other primitive values be. Hence I don't think there is an issue. 

@lukewagner, why would other primitive types require autoboxing either? I think you can treat any tagged pointer as a reference without problem.

Reply:
@rossberg To make sure I understand: are you talking about being able to stick, e.g., a bool into an `anyref` (without autoboxing)?  Also `number`?  Is `anyref` just `any` then?  Is your underlying interest `intref`s here?

Reply:
@lukewagner, yes, there is no technical reason AFAICS why we cannot allow that. Simply view all JS values as reference types. Some of them may be unboxed/flattened in practice, but that is an implementation detail that we can leave entirely transparent to Wasm, because it can never tell the difference.

Intrefs will certainly be in a similar category, but here the motivation is (a) simplicity, and (b) completeness. I think you want all JS value to be able to round-trip through Wasm without having to distinguish them. That makes it much easier to implement generic JS data structures in Wasm.

Reply:
I certainly see the value of round-tripping any JS value (and `any` in general).  When it comes to the question of whether to NaN-box (esp on 32-bit), I think there might be perf benefits to having a more restricted union (not including `f64`).  Probably a good general CG discussion item.